### PR TITLE
Update the adapter for preCICE v3 and add test cases for Elmer-OpenFOAM coupling

### LINKED ISSUE #3: Adapter is not yet ported to preCICE v3
I followed the isntructions of how to compile it. However, I faced an issue with it. Have you seen this error?

`../Coupler_Solver.so: undefined symbol: precicef_is_action_required_

### PR DESCRIPTION
**Pull request by tapegoji**
Description

- Ported Elmer adapter from preCICE v2 to v3.
- Added comprehensive test cases for coupling Elmer with OpenFOAM.
- Ensured compatibility and improved integration with the latest preCICE version.

Changes:

- Updated adapter files to align with preCICE v3 API.
- Included new test cases for Elmer to OpenFOAM coupling. This change required hard-coded naming translation in the Coupler_Solver.F90 code. 
- Enhanced documentation to reflect the changes and new usage instructions.

Testing:

- Successfully ran all new test cases to verify the adapter's functionality with preCICE v3 and OpenFOAM.
- Ensured backward compatibility and verified no breaking changes with existing setups.

Additional Notes:

- Please review the updated documentation and test cases.
- Looking forward to feedback and suggestions for further improvements.

### CODE REVIEW

**Review by uekerman — COMMENTED** [review]
Thanks a lot already! As this change is a bit larger, please be patient with the review :grin: 

Two things already to simplify the review. Please
- Remove the `precice-profiling` folders
- Strip the `tools` to what you actually need

**Comment by tapegoji** [comment]
I have removed the precice-profiling and the extra files in the tools folder. Let me know if you need anything else. I am happy to help make this a great tool. 
Regards,

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 1]
```diff
@@ -23,21 +23,26 @@ SUBROUTINE Print(dataName,mesh,BCPerm,CoordVals)
         TYPE(Mesh_t), POINTER               :: mesh
         !------------------------Data Arrays----------------------------------------------
         REAL(KIND=dp), POINTER              :: CoordVals(:)
-        INTEGER, POINTER                    :: BCPerm(:)
+        INTEGER, POINTER                    :: BoundaryPerm(:)
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]: What was the thought behind this name change? Was it just to make it more clear as to what the variable is?

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 2]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
     Real(KIND=dp) :: Time
 
     !--------------------------Precice-Variables-------------------------------------
     !-------------------------Strings----------------------------------------------
-    CHARACTER(LEN=MAX_NAME_LEN)         :: participantName, meshName, configPath, readDataName,writeDataName
-    ! CHARACTER(LEN=MAX_NAME_LEN)         :: writeInitialData, readItCheckp, writeItCheckp ! ?? Do not know
-    CHARACTER*50                        :: writeInitialData, readItCheckp, writeItCheckp
+    CHARACTER(LEN=MAX_NAME_LEN)         :: config
+    CHARACTER(LEN=MAX_NAME_LEN)         :: participantName, meshName
+    CHARACTER(LEN=MAX_NAME_LEN)         :: readDataName,writeDataName
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
    CHARACTER(LEN=MAX_NAME_LEN)         :: readDataName, writeDataName
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 3]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
-    !--------------------------Initialize-Start-------------------------------------------
-    CALL Info(''//achar(27)//'[31mCouplerSolver ', 'Transfering results between different software ')
 
+    !--------------------------Initialize-Start-------------------------------------------
     Simulation => GetSimulation()
     Mesh => Solver % Mesh
     solverParams => GetSolverParams()
-
+    meshDim = Mesh % MaxDim
+    ! PRINT *, 'Mesh dimension is:', meshDim
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
    meshDim = Mesh % MaxDim
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 4]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
-        !     BCPerm, vertexSize )
-        CALL MakePermUsingMask( Model, Solver, Mesh, MaskName, .TRUE., &
-            BCPerm, vertexSize )
-        CALL Info('CouplerSolver','Number of nodes at interface:'//TRIM(I2S(vertexSize)))
-
         
-
-        ALLOCATE( CoordVals(2*vertexSize) )
-        ALLOCATE(vertexIDs(vertexSize)) 
+        !-----------------Convert to Precice Naming Convention
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        !-----------------Convert to preCICE Naming Conventions
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 5]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
+        IF (meshName == 'fluid-mesh') THEN
+            meshName = 'Fluid-Mesh'
+        END IF
+        !---------------------------------------------------------------------
+
+        !-----------------Get Config Path-------------------------------------
+        config = GetString( Simulation, 'configPath', Found )
+
+        Print *, TRIM(BoundaryName)," ",TRIM(participantName)," ",TRIM(meshName)," ",TRIM(config)
+         !-----------Identify Vertex on Coupling Interface & Save Coordinates--------------------
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        Print *, TRIM(BoundaryName)," ",TRIM(participantName)," ",TRIM(meshName)," ",TRIM(config)
        
         !-----------Identify Vertex on Coupling Interface & Save Coordinates--------------------
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 6]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
-        CALL Print(writeDataName,mesh ,BCPerm,CoordVals)
+        ! !-----------Identify read and write Variables and Create it if it does not exist--------------------
+        CALL CreateVariable(readDataName,'readDataName',mesh,BoundaryPerm,Solver,solverParams)
+        CALL CreateVariable(writeDataName,'writeDataName',mesh,BoundaryPerm,Solver,solverParams)
+        !-----------------------------------------------------------------------------------------
 
-        ! !------------------------------------------------------------------------------
-        ! !---------------Initializing Precice------------------------------------------
+        ! !---------------Initializing Precice------------------------------------------ 
+        CALL Info('CouplerSolver','Initializing Precice')
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        CALL Info('CouplerSolver','Initializing preCICE')     
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 7]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
-        CALL precicef_get_dims(dimensions)
-        CALL precicef_get_mesh_id(meshName, meshID)
-        CALL precicef_set_vertices(meshID, vertexSize, CoordVals, vertexIDs)
-
-        CALL precicef_get_data_id(readDataName,meshID,readDataID)
-        CALL precicef_get_data_id(writeDataName,meshID,writeDataID)
-
-        ALLOCATE(readData(VertexSize))
-        ALLOCATE(writeData(VertexSize))
+        CALL Info('CouplerSolver','Setting up mesh in Precice')
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        CALL Info('CouplerSolver','Setting up mesh in preCICE')
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 8]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
         IF (bool.EQ.1) THEN
-            CALL Info('CouplerSolver','Writing Initial Data')
-            CALL CopyWriteData(writeDataName,mesh,BCPerm,writeData)
-            CALL precicef_write_bsdata(writeDataID, vertexSize, vertexIDs, writeData)
-            CALL precicef_mark_action_fulfilled(writeInitialData)
+            WRITE (*,*) 'TODO: Provide initial data if needed'
+            CALL CopyWriteData(writeDataName,mesh,BoundaryPerm,writeData)
+            IF (writeDataName == 'temperature') THEN
+                CALL precicef_write_data(meshName, 'Temperature', BoundaryNodes, vertexIDs, writeData)
+            ELSE IF (writeDataName == 'temperature flux_abs') THEN
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]: Why is the variable called `temperature flux_abs` and not just `flux_abs`?

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 9]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
-        CALL Info('CouplerSolver','Copy Read Data to Variable')
-        CALL CopyReadData(readDataName,mesh,BCPerm,readData)
-
-        CALL Info('CouplerSolver','Printing read Data')
-        CALL Print(readDataName,mesh ,BCPerm,CoordVals)
         
-
-        CALL Info('CouplerSolver','Printing write Data')
-        CALL Print(writeDataName,mesh ,BCPerm,CoordVals)
+        CALL Info('CouplerSolver ', 'Readinging the data from Precice')
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        CALL Info('CouplerSolver ', 'Reading data from preCICE')    
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 10]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
-
-        CALL Info('CouplerSolver','Printing write Data')
-        CALL Print(writeDataName,mesh ,BCPerm,CoordVals)
+        CALL Info('CouplerSolver ', 'Readinging the data from Precice')    
+        CALL precicef_get_max_time_step_size(dt)
         
+        !-------------------Sticking Precice Naming Convention-------------------------------------
+        IF (readDataName == 'temperature') THEN
+            CALL precicef_read_data(meshName, 'Temperature', BoundaryNodes, vertexIDs, dt, readData)
+        ELSE IF (readDataName == 'temperature flux_abs') THEN
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]: Same comment as above.

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 11]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
+            CALL precicef_read_data(meshName, 'Heat-Flux', BoundaryNodes, vertexIDs, dt, readData)
+        ELSE
+            CALL precicef_read_data(meshName, readDataName, BoundaryNodes, vertexIDs, dt, readData)
+        END IF
+
+        CALL CopyReadData(readDataName,mesh,BoundaryPerm,readData)
+        CALL Info('CouplerSolver','Printing the read data(From Precice to Elmer)')
+        ! CALL Print(readDataName,mesh ,BoundaryPerm,CoordVals)
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 12]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
         itask = 3
     case(3)
-        !-------------------Copy Write values from Variable to buffer---------------------
-        
-        CALL Info('CouplerSolver','Copy Write Data to Variable')
-        CALL CopyWriteData(writeDataName,mesh,BCPerm,writeData)
-
-        CALL Info('CouplerSolver','Writing Data')
-        CALL precicef_write_bsdata(writeDataID, vertexSize, vertexIDs, writeData)
+        !-------------------Copy Write values from Variable to buffer----------------------------
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion

        !-------------------Copy Write values from Variable to buffer----------------------------
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 13]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
+        CALL CopyWriteData(writeDataName,mesh,BoundaryPerm,writeData)
+        CALL Info('CouplerSolver','Printing the write data (From Elmer to Precice)')
+        ! CALL Print(writeDataName,mesh ,BoundaryPerm,CoordVals)
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 14]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
+        END IF
 
-        CALL Info('CouplerSolver','Printing write Data')
-        CALL Print(writeDataName,mesh ,BCPerm,CoordVals)
         
-        CALL Info('CouplerSolver','Printing read Data')
-        CALL Print(readDataName,mesh ,BCPerm,CoordVals)
-
-        !--------------------Advance time loop---------------------------------------
+        !-------------------Advance Precice-------------------------------------------------------
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        !-------------------Advance preCICE-------------------------------------------------------
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 15]
```diff
@@ -182,288 +186,228 @@ SUBROUTINE CouplerSolver( Model,Solver,dt,TransientSimulation)
...
         ELSE
             itask = 2
         END IF
-          
-    case(4)    
-        !----------------------------------------Finailize--------------------------------------
+        !-----------------------------------------------------------------------------------------
+
+    case(4)
         CALL Info('CouplerSolver','Precice Finalize')
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        CALL Info('CouplerSolver','preCICE Finalize')
```

File: Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log:1]: This log file can be removed.

File: Flow_Over_Heated_Plate/README.md
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/README.md:1]: This README would be better suited on main folder level

File: Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log:1]: This log file can be removed.

File: Flow_Over_Heated_Plate/precice-config.xml [hunk 16]
```diff
@@ -1,59 +1,57 @@
...
+    <read-data name="Heat-Flux" mesh="Fluid-Mesh"/>
+    <mapping:nearest-neighbor
+      direction="read"
+      from="Solid-Mesh"
+      to="Fluid-Mesh"
+      constraint="consistent" />
+  </participant>
+
+  <m2n:sockets acceptor="Solid" connector="Fluid" exchange-directory=".."/>
+
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/precice-config.xml]:
```suggestion
```

**Review by IshaanDesai — CHANGES_REQUESTED** [review]
@tapegoji thanks a lot for this contribution! Great work already :clap: Please have a look at my review comments. I did not run the test case and also did not have a look at the OpenFOAM things. After we do one iteration of the review, I can try running things. We can then finalize this pull request.

File: Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log:1]: Yea I agree

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 (hunk 8 above)
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Coupler_Solver.F90]: Thad the name of Elmer variable. Not sure with just flux_abs it works. For now leave it to this. I have a new change that is coming and that will change this to magnitude. Elmer has different flux for different physics. By saying temperature flux it means heat flux.

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 (hunk 10 above)
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Coupler_Solver.F90]: Ok

**Comment by IshaanDesai** [comment]
@tapegoji please let me know when the pull request is ready to be reviewed again. No rush, but I will wait for your ping to have a look again.

**Comment by tapegoji** [comment]
@IshaanDesai, I am a bit confused what I need to do. can you explain what I should do. 
Thank you. I am a bit new and not too familiar with the process.

**Comment by tapegoji** [comment]
I have applied all the changes. let me know if I am missing anything.

**Comment by tapegoji** [comment]
Any update on this?

**Comment by IshaanDesai** [comment]
> Any update on this?

I'll review this week :+1: apologies for the delay, some other urgent things took time.

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 [hunk 17]
```diff
@@ -97,37 +102,36 @@ SUBROUTINE CreateVariable(dataName,dataType,mesh,BCPerm,Solver,solverParams)
...
-            IF(.NOT. Found ) Dofs = 1
-            CALL VariableAddVector( mesh % Variables, mesh, Solver, dataName, Dofs, &
-                Perm = BCPerm, Secondary = .TRUE. )
-            dataVariable => VariableGet( mesh % Variables, dataName ) 
+            CALL FATAL('CouplerSolver', 'Variable does not exist : ' // TRIM(dataName) )
+            ! Dofs = ListGetInteger( solverParams,'Field Dofs',Found )
+            ! IF(.NOT. Found ) Dofs = 1
+            ! CALL VariableAddVector( mesh % Variables, mesh, Solver, dataName, Dofs, &
+            !     Perm = BoundaryPerm, Secondary = .TRUE. )
+            ! dataVariable => VariableGet( mesh % Variables, dataName )
```
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]: Can we remove these lines of code?

File: Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log:1]: Please delete this file.

File: Flow_Over_Heated_Plate/README.md
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/README.md:1]: Why is this README file duplicated? Unless absolutely required, let us only keep the README at the base adapter level.

File: Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log:1]: Please delete this file.

**Review by IshaanDesai — APPROVED** [review]
We are close to merging this. I have added some last comments. For the flow over heated plate tutorial, I see that you have copied over the OpenFOAM files from the [precice/tutorials](https://github.com/precice/tutorials/tree/develop) repository. Instead of copying the files here, a better idea would be to write a small README file here which describes how to copy Elmer files to the tutorials repository and run the tutorial there. This would avoid duplication. If it is not clear how to do this, I or someone else can do this.

Thanks once more for your contribution! Once you address the comments above, we can finalize this.

File: Flow_Over_Heated_Plate/Coupler_Solver.F90 (hunk 17 above)
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Coupler_Solver.F90]: Yes, Let's delete them. they were in the original one and I did not know why.

File: Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log:1]: Done!

File: Flow_Over_Heated_Plate/README.md
Comment (tapegoji) [reply Flow_Over_Heated_Plate/README.md:1]: I agree, Ill remove it

File: Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log:1]: Done

**Comment by IshaanDesai** [comment]
@tapegoji thanks for making the final changes. Adapter-wise we are ready to merge this. The next step would be to move the files relevant to the flow-over-heated-plate tutorial to the [precice/tutorials](https://github.com/precice/tutorials/tree/develop) repository. I can do this in the next days, and then we can test if everything works. Once we are sure, we can merge this pull request, and also merge Elmer as a new participant to the tutorial.

**Comment by IshaanDesai** [comment]
The Elmer fluid and solid participants for the flow-over-heated-plate tutorial are being added to the appropriate folder in the precice/tutorials repository via: https://github.com/precice/tutorials/pull/565

The cases, and also the OpenFOAM cases copied over for testing have now been deleted. The reference problem has also been removed. The `tools/` folder used for cleanups has also been removed.

I will continue cleaning up this pull request and finalizing the one mentioned above.

File: .gitignore
Comment (IshaanDesai) [inline .gitignore:1]: Lets remove these lines as they are no longer required.

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90:1]: I am curious as to why this file is necessary when there is a very similar file `Adapter/Coupler_Solver.F90`. I would suggest moving this file to the `Adapter/` folder, and replacing the older file.

File: Flow_Over_Heated_Plate3D/Reference_Problem/Reference_Problem_Mesh.msh
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate3D/Reference_Problem/Reference_Problem_Mesh.msh:1]: I think we can remove the reference problem after the testing is done. I already removed the reference problem for the 2D case.

**Review by IshaanDesai — COMMENTED** [review]
Here are somethings I noticed while moving the case files to the `precice/tutorials` repository.

**Comment by tapegoji** [comment]
I am good with all the changes. Please go ahead.

**Review by tapegoji — COMMENTED** [review]
sounds good to me

**Comment by IshaanDesai** [comment]
@tapegoji I opened a pull request in the tutorials repository to move the flow over heated plate case files to the respective folder: https://github.com/precice/tutorials/pull/565 I tried running the case there, with both `solid-elmer` and `fluid-elmer`, but the solver did not run as the mesh file was not detected. Can you please have a look and try to run the cases there? I am a bit short on time until the end of next week due to the [upcoming preCICE workshop](https://precice.org/precice-workshop-2024.html).

**Comment by tapegoji** [comment]
did you use the mesh from OpenFoam or did you use the mesh that I provided?

**Comment by IshaanDesai** [comment]
> did you use the mesh from OpenFoam or did you use the mesh that I provided?

I used the mesh which was in the folder of the case in this pull request. Let us continue the discussion in https://github.com/precice/tutorials/pull/565
I am sure it is just something minor. Once the case is running, we can merge everything. Thanks!

**Comment by IshaanDesai** [comment]
Merging this PR and continuing testing https://github.com/precice/tutorials/pull/565 for correctness. The changes in this PR make sense. Minor documentation polishing still needs to be done, but this can also be done on the main branch directly.

**Pull request merged**